File: api/collect.js
import { Redis } from "@upstash/redis";
import Parser from "rss-parser";
import { Resend } from "resend";

// ---- clients ----
const redis = new Redis({
  url: process.env.STORAGE_KV_REST_API_URL,
  token: process.env.STORAGE_KV_REST_API_TOKEN
});

// Enable YouTube/media fields & add requestOptions for UA
const parser = new Parser({
  customFields: {
    item: [
      ['media:group', 'media', { keepArray: false }],
      ['media:description', 'mediaDescription'],
      ['media:content', 'mediaContent', { keepArray: false }],
      ['media:thumbnail', 'mediaThumb', { keepArray: false }],
    ]
  },
  requestOptions: {
    headers: {
      'User-Agent':
        'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 ' +
        '(KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
      'Accept': 'application/rss+xml, application/atom+xml, application/xml;q=0.9, */*;q=0.8',
    },
    timeout: 10000
  }
});

const resend = process.env.RESEND_API_KEY ? new Resend(process.env.RESEND_API_KEY) : null;

// ---- storage keys ----
const ZSET = "mentions:z";
const SEEN_ID = "mentions:seen";
const SEEN_LINK = "mentions:seen:canon";
const MAX_MENTIONS = 5000;

// ---- config ----
// Support both comma and semicolon delimiters for RSS_FEEDS
const RSS_FEEDS = (process.env.RSS_FEEDS || "").split(/[,;]/).map(s => s.trim()).filter(Boolean);

// ---- helpers ----
function normalizeUrl(u) {
  try {
    const url = new URL(u);
    url.hash = "";
    ["utm_source","utm_medium","utm_campaign","utm_term","utm_content","utm_id",
     "mc_cid","mc_eid","ref","fbclid","gclid","igshid"].forEach(p => url.searchParams.delete(p));
    if ([...url.searchParams.keys()].length === 0) url.search = "";
    url.hostname = url.hostname.toLowerCase();
    let s = url.toString();
    if (s.endsWith("/")) s = s.slice(0, -1);
    return s;
  } catch {
    return (u || "").trim();
  }
}
function hostOf(u) { try { return new URL(u).hostname.toLowerCase(); } catch { return ""; } }
function normalizeHost(h) { return (h || "").toLowerCase().replace(/^www\./, "").replace(/^amp\./, ""); }
function unwrapGoogleAlert(u) {
  try {
    const url = new URL(u);
    if (url.hostname.endsWith("google.com") && url.pathname === "/url") {
      return url.searchParams.get("q") || url.searchParams.get("url") || u;
    }
    return u;
  } catch { return u; }
}
function displaySource(link, fallback) { const h = normalizeHost(hostOf(link)); return h || (fallback || ""); }
function buildYouTubeWatchUrl(s) {
  s = (s || "").trim();
  if (/^https?:\/\//i.test(s)) return s;
  if (/^[A-Za-z0-9_-]{11}$/.test(s)) return `https://www.youtube.com/watch?v=${s}`;
  return s;
}
function extractItemLink(e) {
  let raw =
    (e.link && typeof e.link === "object" && e.link.href) ? e.link.href :
    (Array.isArray(e.link) && e.link[0]?.href)            ? e.link[0].href :
    (e.links && e.links[0]?.href)                         ? e.links[0].href :
    (typeof e.link === "string" ? e.link : "") ||
    (typeof e.id === "string" ? e.id : "");

  raw = unwrapGoogleAlert(raw);

  const ytId =
    e["yt:videoId"] ||
    e.videoId ||
    (typeof e.id === "string" && e.id.startsWith("yt:video:") ? e.id.split("yt:video:")[1] : "");

  if (!/^https?:\/\//i.test(raw) && ytId) raw = buildYouTubeWatchUrl(ytId);
  else {
    const h = hostOf(raw);
    if (h.includes("youtube.com") || h.includes("youtu.be")) raw = buildYouTubeWatchUrl(raw);
  }
  return (raw || "").trim();
}
function idFromCanonical(c) { let h=0; for (let i=0;i<c.length;i++) h=(h*31+c.charCodeAt(i))>>>0; return `m_${h.toString(16)}`; }
function toEpoch(d){ const t=Date.parse(d); return Number.isFinite(t)?Math.floor(t/1000):Math.floor(Date.now()/1000); }
const ENABLE_SENTIMENT = (process.env.ENABLE_SENTIMENT || "").toLowerCase() === "true";
const POS = ["win","surge","rally","gain","positive","bull","record","secure","approve","partnership"];
const NEG = ["hack","breach","lawsuit","fine","down","drop","negative","bear","investigate","halt","outage","delay","ban"];
function sentimentScore(text){
  const t = (text||"").toLowerCase();
  let s = 0;
  for (const w of POS) if (t.includes(w)) s += 1;
  for (const w of NEG) if (t.includes(w)) s -= 1;
  return s;
}
async function sendEmail(m){
  if(!resend || !process.env.ALERT_EMAIL_FROM || !process.env.ALERT_EMAIL_TO) return;
  const to = process.env.ALERT_EMAIL_TO.split(",").map(s=>s.trim()).filter(Boolean);
  if(!to.length) return;
  await resend.emails.send({
    from: process.env.ALERT_EMAIL_FROM,
    to,
    subject: `[URGENT] ${m.title}`,
    html: `<p><b>${m.title}</b></p>
           <p>Source: ${m.source} · ${m.published}</p>
           <p>Section: ${m.section}</p>
           <p><a href="${m.link}">Open article</a></p>`
  });
}

// ---- handler ----
export default async function handler(req, res) {
  try {
    let found = 0, stored = 0, emailed = 0, errors = [];

    // Check if RSS feeds are configured
    if (!RSS_FEEDS.length) {
      console.log('RSS_FEEDS not configured - skipping RSS collection');
      res.status(200).json({
        ok: true,
        message: "RSS collection disabled - no feeds configured",
        found: 0,
        stored: 0,
        emailed: 0,
        errors: [],
        rss_disabled: true,
        generated_at: new Date().toISOString()
      });
      return;
    }

    // No keyword filtering - collect all articles
    console.log(`RSS collection starting: ${RSS_FEEDS.length} feeds, no keyword filtering`);

    for (const url of RSS_FEEDS) {
      try {
        const feed = await parser.parseURL(url);
        const feedTitle = feed?.title || url;

        for (const e of feed?.items || []) {
          const title = (e.title || "").trim();
          const ytDesc = e.mediaDescription || e?.media?.description || e?.mediaContent?.description || "";
          const sum = ytDesc || e.contentSnippet || e.content || e.summary || "";
          const link = extractItemLink(e);

          // No keyword filtering - accept all articles
          const canon = normalizeUrl(link || title);
          if (!canon) continue;

          const addCanon = await redis.sadd(SEEN_LINK, canon);
          if (addCanon !== 1) continue;

          const mid = idFromCanonical(canon);
          await redis.sadd(SEEN_ID, mid);

          const ts = toEpoch(e.isoDate || e.pubDate || e.published || e.updated);

          const m = {
            id: mid,
            canon,
            section: "Google Alerts",
            title: title || "(untitled)",
            link,
            source: displaySource(link, feedTitle),
            summary: sum,
            origin: "google_alerts",
            published_ts: ts,
            published: new Date(ts * 1000).toISOString()
          };

          if (ENABLE_SENTIMENT) m.sentiment = sentimentScore(`${title} ${sum}`);
          await redis.zadd(ZSET, { score: ts, member: JSON.stringify(m) });

          const count = await redis.zcard(ZSET);
          if (count > MAX_MENTIONS) await redis.zremrangebyrank(ZSET, 0, count - MAX_MENTIONS - 1);

          found++; stored++;
        }
      } catch (err) {
        errors.push({ url, error: err?.message || String(err) });
      }
    }
    res.status(200).json({ ok:true, feeds: RSS_FEEDS.length, found, stored, emailed, errors });
  } catch (e) {
    res.status(500).json({ ok:false, error:`collect failed: ${e?.message || e}` });
  }
}

File: api/collect.py
import os, time, json
from datetime import datetime
from typing import List, Dict
from http.server import BaseHTTPRequestHandler
import feedparser
from dateutil import parser as dtp
from upstash_redis import Redis
from resend import Emails

REDIS = Redis(url=os.getenv("KV_REST_API_URL"), token=os.getenv("KV_REST_API_TOKEN"))
RSS_FEEDS: List[str] = [u.strip() for u in (os.getenv("RSS_FEEDS") or "").split(",") if u.strip()]
KEYWORDS:  List[str] = [k.strip().lower() for k in (os.getenv("KEYWORDS") or "").split(",") if k.strip()]
URGENT:    List[str] = [k.strip().lower() for k in (os.getenv("ALERT_KEYWORDS_URGENT") or "").split(",") if k.strip()]

ZSET = "mentions:z"
SEEN = "mentions:seen"
MAX_MENTIONS = 5000

RESEND_API_KEY   = os.getenv("RESEND_API_KEY")
ALERT_EMAIL_FROM = os.getenv("ALERT_EMAIL_FROM")
ALERT_EMAIL_TO   = [e.strip() for e in (os.getenv("ALERT_EMAIL_TO") or "").split(",") if e.strip()]

def _json(self, obj, status=200):
    data = json.dumps(obj, separators=(",", ":"), ensure_ascii=False).encode("utf-8")
    self.send_response(status)
    self.send_header("Content-Type", "application/json; charset=utf-8")
    self.send_header("Content-Length", str(len(data)))
    self.end_headers()
    self.wfile.write(data)

def _now() -> int: return int(time.time())

def _pub_ts(e) -> int:
    for field in ("published", "updated", "pubDate"):
        v = getattr(e, field, None) or e.get(field)
        if v:
            try: return int(dtp.parse(v).timestamp())
            except Exception: pass
    return _now()

def _match(t: str) -> List[str]:
    t = (t or "").lower()
    return [k for k in KEYWORDS if k in t]

def _urgent(matched: List[str]) -> bool:
    return bool(set(matched) & set(URGENT)) if URGENT else False

def _id(link: str, title: str) -> str:
    from hashlib import sha256
    return sha256((link or title or "").encode("utf-8")).hexdigest()

def _store(m: Dict):
    REDIS.zadd(ZSET, {json.dumps(m, separators=(",", ":")): m["published_ts"]})
    REDIS.zremrangebyrank(ZSET, 0, -MAX_MENTIONS-1)

def _email(m: Dict):
    if not (RESEND_API_KEY and ALERT_EMAIL_FROM and ALERT_EMAIL_TO): return
    Emails.api_key = RESEND_API_KEY
    Emails.send({
        "from": ALERT_EMAIL_FROM,
        "to": ALERT_EMAIL_TO,
        "subject": f"[URGENT] {m['title']}",
        "html": (
            f"<p><b>{m['title']}</b></p>"
            f"<p>Source: {m['source']} · {m['published']}</p>"
            f"<p>Keywords: {', '.join(m['matched'])}</p>"
            f"<p><a href='{m['link']}'>Open</a></p>"
        )
    })

class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        try:
            if not (RSS_FEEDS and KEYWORDS):
                _json(self, {"ok": False, "error": "Missing RSS_FEEDS or KEYWORDS"}, 400); return

            found = stored = emailed = 0

            for url in RSS_FEEDS:
                feed = feedparser.parse(url)
                source = (getattr(feed, "feed", {}) or {}).get("title") or url

                for e in getattr(feed, "entries", []):
                    title = (e.get("title") or "").strip()
                    link  = (e.get("link")  or "").strip()
                    matched = _match(f"{title}\n{e.get('summary','')}\n{link}")
                    if not matched: continue

                    mid = _id(link, title)
                    if REDIS.sadd(SEEN, mid) != 1:  # 1=new, 0=seen
                        continue

                    ts = _pub_ts(e)
                    m = {
                        "id": mid,
                        "title": title or "(untitled)",
                        "link": link,
                        "source": source,
                        "matched": matched,
                        "published_ts": ts,
                        "published": datetime.utcfromtimestamp(ts).isoformat() + "Z"
                    }
                    _store(m)
                    stored += 1
                    if _urgent(matched): _email(m); emailed += 1
                    found += 1

            _json(self, {"ok": True, "feeds": len(RSS_FEEDS), "found": found, "stored": stored, "emailed": emailed}, 200)
        except Exception as e:
            _json(self, {"ok": False, "error": f"collect failed: {type(e).__name__}: {e}"}, 500)
